akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -548,3 +545,210 @@ final private[stream] class LazySink[T, M](sinkFactory: T ⇒ Future[Sink[T, M]]
     (stageLogic, promise.future)
   }
 }
+
+/**
+ * INTERNAL API
+ */
+final class FoldResourceSink[T, S](
+  open:      () ⇒ S,
+  writeData: (S, T) ⇒ Unit,
```

Comment:
This stage is intended to be used with an imperative resource like file, database, network socket. `S` represents a file descriptor, database connection or socket descriptor. So far I can't imagine changing `S` by writing data.

(I suppose it would be `(S, T) => IO[Unit]` if we were in a purely functional world.)